cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -353,6 +353,332 @@ struct IfElseFunctor<Type, enable_if_number<Type>> {
   }
 };
 
+template <typename Type>
+struct IfElseFunctor<Type, enable_if_base_binary<Type>> {
```

Comment:
It looks like none of these cases handle the validity buffer yet?

Reply:
Also, you don't have to pursue this, but it feels like you can consolidate the four cases by templating a core loop on a pair of accessor structs, which have a method to copy a block of entries into the output data buffer and into the output offsets buffer.

Reply:
validity buffer is handles prior to calling this actually. 
https://github.com/apache/arrow/blob/2e5af682ed8afb8867b97c8f142710f1b0513b87/cpp/src/arrow/compute/kernels/scalar_if_else.cc#L849

Reply:
> Also, you don't have to pursue this, but it feels like you can consolidate the four cases by templating a core loop on a pair of accessor structs, which have a method to copy a block of entries into the output data buffer and into the output offsets buffer.

Would you have some example to this? I'd love to pursue something like that. Because there's a lot of boiler plate code here.

Reply:
Oh, right, you're right, my bad.

I haven't fully thought this out but I was thinking something like

```
struct UnwrapScalar {
    const Scalar& scalar;

    void WriteOffsets(uint8_t* out_offsets, int64_t offset, int64_t length) {
        // simple loop...
    }
}

struct UnwrapArray {
    const ArrayData& array;

    void WriteOffsets(uint8_t* out_offsets, int64_t offset, int64_t length) {
        // std::transform...
    }
}

template <typename UnwrapLeft, typename UnwrapRight>
Status CallImpl(const ArrayData& cond, const UnwrapLeft& lhs, const UnwrapRight& rhs) {
    ...
    while (...) {
        if (block.AllSet()) {
            lhs.WriteOffsets(...);
            lhs.CopyData(...);
        }
    }
}

Status Call(const ArrayData& cond, const Scalar& lhs, const ArrayData& rhs) {
    return CallImpl(cond, UnwrapScalar{lhs}, UnwrapArray{rhs});
}

// ... more overloads
```

which I think would generate similar code while deduplicating the cases (somewhat).